0006-ffmul-revised.md

```diff
@@ -509,10 +520,1725 @@ In summary, we chose the gate design presented in this RFC because:
+19. $v_1 = v_{1,0} + v_{1,12}*2^{12} + v_{1,24}*2^{24} + v_{1,36}*2^{36} + v_{1,48}*2^{48} + v_{1,60}*2^{60} + v_{1,72}*2^{72} + v_{1,84}*2^{84} + v_{1,86}*2^{86} + v_{1,88}*2^{88} + v_{1,90}*2^{90}$
+20. $2^{\ell} \cdot v_1 = v_0 + p_{11} + p_2 - r_2$
+21. $a_n \cdot b_n + q_n \cdot f'_n = r_n$
+22. $q'_2 = q_2 + 2^\ell - f_2 - 1$
+23. $q'_2 \in [0, 2^\ell]$ `multi-range-check`
+
+## Appendix J: Constraints
+
+These checks can be condensed into the minimal number of constraints as follows.
+
+Note that (4-10) are carried out by the lookup argument.
+
+First, we have the range-check corresponding to (1) as a degree-4 constraint
+
+**C1:** $p_{111} \cdot (p_{111} - 1) \cdot (p_{111} - 2) \cdot (p_{111} - 3)$
+
+Now we have the range-check corresponding to (15) as another degree-4 constraint
+
+**C2:** $v_0 \cdot (v_0 - 1) \cdot (v_0 - 2) \cdot (v_0 - 3)$
+
+Next (16) and (17) can be combined into
+
+**C3:** $2^{\ell} \cdot (2^{\ell} \cdot p_{111} + p_{110}) + p_{10} = p_1$
+
+Next we have check (18)
+
+**C4:** $2^{2\ell} \cdot v_0 = p_0 + 2^{\ell} \cdot p_{10} - r_{01}$
+
+Next, for our use of the CRT, we must constrain that $a \cdot b = q \cdot f + r \mod n$.  Thus, check (21) is
+
+**C5:** $a_n \cdot b_n + q_n \cdot f'_n = r_n$
+
+**C6:** (11) $v_{1,84}$ (2-bit check)
+
+$v_{1,84} \cdot (v_{1,84} - 1) \cdot (v_{1,84} - 2) \cdot (v_{1,84} - 3)$
+
+**C7:** (12) $v_{1,86}$ (2-bit check)
+
+$v_{1,86} \cdot (v_{1,86} - 1) \cdot (v_{1,86} - 2) \cdot (v_{1,86} - 3)$
+
+**C8:** (13) $v_{1,88}$ (2-bit check)
+
+$v_{1,88} \cdot (v_{1,88} - 1) \cdot (v_{1,88} - 2) \cdot (v_{1,88} - 3)$
+
+**C9:** (14) $v_{1,90}$ (1-bit check)
+
+$v_{1,90} \cdot (v_{1,90} - 1) $
+
+Now checks (19) and (20) can be combined into
+
+**C10:**  $2^{\ell} \cdot (v_{1,0} + v_{1,12}*2^{12} + v_{1,24}*2^{24} + v_{1,36}*2^{36} + v_{1,48}*2^{48} + v_{1,60}*2^{60} + v_{1,72}*2^{72} + v_{1,84}*2^{84} + v_{1,86}*2^{86} + v_{1,88}*2^{88} + v_{1,90}*2^{90}) = p_2 + p_{11} + v_0 - r_2$
+
+Check (22) is correctly computed
+
+**C11:** $q'_2 - q_2 + 2^\ell + f_2 + 1$.
+
+**MRC:** `multi-range-check` $q'_2, p_{10}, p_{110}$
+
+**MRC:** `1.33 multi-range-check` $r_2`
+
+**MRC:** `compact-multi-range-check` $r_01, r_2$
+
+**MRC:** `multi-range-check` $q_0, q_1, q_2$
+
+## Appendix K: External checks
+
+The following checks must be done with other gates to assure the soundness of the foreign field multiplication
+
+- Range check input
+    - `multi-range-check` $a$
+    - `multi-range-check` $b$
+- Compute and constrain high bounds of inputs $a'_2$ and $b'_2$
+    - `Generic` $a_2 + (2^\ell - f_2 -1) = a'_2$ and $b_2 + (2^\ell - f_2 -1) = b'_2$
+- If $r$ needs to be canonical ($<f$ compute a full bound check: 
+    - `ForeignFieldAdd` $r + 2^t = 1 \cdot f + r'$
+    - `multi-range-check` $r'$
+
+Copy constraints must connect the above witness cells to their respective input cells within the corresponding external check gates witnesses.
+
+## Appendix L: Soundness
```

Comment:
In my opinion, we don't need this section, because the same proof in a more suitable form is already contained in the main section. I think we should only keep the last part, "necessity of range checks"

Reply:
Yeah, probably I will just link it inside "Prior art", and just extract the necessity of range checks section, which is mainly the reason why we are switching the design.